cgo/arith.c

```diff
@@ -15,3 +15,145 @@
  */
 
 #include "mo_impl.h"
+
+/* 
+ * Signed int add with overflow check.
+ *
+ * The test checks if rt[i] and tmpx have different sign, and rt[i]
+ * and tmpy have different sign, therefore, tmpx and tmpy have same 
+ * sign but the result has different sign, therefore, it is an overflow.
+ *
+ * DO NOT use __builtin_add_overflow, which gcc cannot vectorize with SIMD
+ * and is actually much much slower.
+ */
+#define ADD_SIGNED_OVFLAG(TGT, A, B)                 \
+    TGT = (A) + (B);                                 \
+    opflag |= ((TGT) ^ (A)) & ((TGT) ^ (B))
+
+#define ADD_SIGNED_OVFLAG_CHECK                      \
+    if (opflag < 0)    {                             \
+        return RC_OUT_OF_RANGE;                      \
+    } else return RC_SUCCESS
+
+
+/*
+ * Unsigned int add with overflow check
+ *
+ * If result is less, we know it wrapped around.
+ */
+#define ADD_UNSIGNED_OVFLAG(TGT, A, B)              \
+    TGT = (A) + (B);                                \
+    if ((TGT) < (A)) {                              \
+        opflag = 1;                                 \
+    } else (void) 0
+
+#define ADD_UNSIGNED_OVFLAG_CHECK                   \
+    if (opflag != 0) {                              \
+        return RC_OUT_OF_RANGE;                     \
+    } else return RC_SUCCESS
+
+/*
+ * Float/Double overflow check.
+ *
+ * At this moment we don't do anything.
+ */
+#define ADD_FLOAT_OVFLAG(TGT, A, B)                 \
+    TGT = (A) + (B)    
+
+#define ADD_FLOAT_OVFLAG_CHECK                      \
+    (void) opflag;                                  \
+    return RC_SUCCESS
+
+
+
+#define MO_ARITH_T(OP, ZT)                                    \
+    ZT *rt = (ZT *) r;                                        \
+    ZT *at = (ZT *) a;                                        \
+    ZT *bt = (ZT *) b;                                        \
+    ZT opflag = 0;                                            \
+    if ((flag & 1) != 0) {                                    \
```

Comment:
Add MACROs for flag's values like 0 -- VECTOR_VECTOR, 1 -- SCALAR_VECTOR, 2 -- VECTOR_SCALAR ?